utils/src/types/lockup_script.rs

```diff
@@ -0,0 +1,231 @@
+            return Ok(DecodeStage { ..*stage });
+        }
+        match self {
+            LockupScript::Unknown => {
+                let tpe = buffer.next_byte().unwrap();
+                let result = LockupScript::from_type(tpe);
+                if result.is_none() {
+                    return Err(DecodeError::InvalidData);
+                }
+                *self = result.unwrap();
+            }
+            _ => (),
+        };
+        match self {
+            LockupScript::P2PKH(hash) => hash.decode(buffer, stage),
+            LockupScript::P2MPKH(hashes) => hashes.decode_children(buffer, stage),
+            LockupScript::P2SH(hash) => hash.decode(buffer, stage),
+            LockupScript::P2C(hash) => hash.decode(buffer, stage),
+            LockupScript::Unknown => Err(DecodeError::InternalError),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    extern crate std;
+
+    use crate::buffer::Buffer;
+    use crate::decode::{new_decoder, Decoder};
+    use crate::types::byte32::tests::gen_bytes;
+    use crate::types::i32::tests::random_usize;
+    use crate::types::u256::tests::hex_to_bytes;
+    use crate::types::{Hash, LockupScript};
+    use crate::TempData;
+    use std::vec;
+
+    #[test]
+    fn test_decode_p2pkh() {
+        for _ in 0..10 {
+            let mut bytes = vec![0u8];
+            let hash_bytes = gen_bytes(32, 32);
+            bytes.extend(&hash_bytes);
+            let lockup_script =
+                LockupScript::P2PKH(Hash::from_bytes(hash_bytes.as_slice().try_into().unwrap()));
+            let mut temp_data = TempData::new();
+
+            {
+                let mut buffer = Buffer::new(&bytes, &mut temp_data).unwrap();
+                let mut decoder = new_decoder::<LockupScript>();
+                let result = decoder.decode(&mut buffer).unwrap();
+                assert_eq!(result, Some(&lockup_script));
+            }
+
+            let mut length: usize = 0;
+            let mut decoder = new_decoder::<LockupScript>();
+
+            while length < bytes.len() {
+                let remain = bytes.len() - length;
+                let size = random_usize(0, remain);
+                let mut buffer =
+                    Buffer::new(&bytes[length..(length + size)], &mut temp_data).unwrap();
+                length += size;
+
+                let result = decoder.decode(&mut buffer).unwrap();
+                if length == bytes.len() {
+                    assert_eq!(result, Some(&lockup_script));
+                    assert!(decoder.stage.is_complete())
+                } else {
+                    assert_eq!(result, None);
+                }
+            }
+        }
+    }
+
+    #[test]
+    fn test_decode_p2mpkh() {
+        let bytes = hex_to_bytes("0103a3cd757be03c7dac8d48bf79e2a7d6e735e018a9c054b99138c7b29738c437ecef51c98556924afa1cd1a8026c3d2d33ee1d491e1fe77c73a75a2d0129f061951dd2aa371711d1faea1c96d395f08eb94de1f388993e8be3f4609dc327ab513a02").unwrap();
+        for _ in 0..10 {
```

Comment:
[Nit] Seems that we don't need to repeat here.

Reply:
Yes, will fix it 👍